certificate_manager: Check that template differs from current cert before rotation

agunnerson-ibm opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

With the current behavior, when kubelet starts, a `templateChanged`
event is always fired off because it only checks if `getLastRequest`
matches `getTemplate`. The last request only exists in memory and thus
is initially `nil` and can't ever match the current template during
startup.

This causes kubelet to request the signing of a new CSR every time it's
restarted. This commit changes the behavior so that `templateChanged` is
only fired off if the currently template doesn't match both the current
certificate and the last template.

**Which issue(s) this PR fixes**:

Fixes #69471

**Release note**:

```release-note
When `--rotate-server-certificates` is enabled, kubelet will no longer request a new certificate on startup if the current certificate on disk is satisfactory.
```

agunnerson-ibm committed:
certificate_manager: Check that template differs from current cert before rotation

With the current behavior, when kubelet starts, a `templateChanged`
event is always fired off because it only checks if `getLastRequest`
matches `getTemplate`. The last request only exists in memory and thus
is initially `nil` and can't ever match the current template during
startup.

This causes kubelet to request the signing of a new CSR every time it's
restarted. This commit changes the behavior so that `templateChanged` is
only fired off if the currently template doesn't match both the current
certificate and the last template.

Fixes #69471

Signed-off-by: Andrew Gunnerson <andrew.gunnerson@us.ibm.com>

fedebongio commented:
/cc @caesarxuchao

liggitt commented:
/ok-to-test

liggitt commented:
there's a gofmt error that needs fixing to make the bot happy:
```
-func (m* manager) certSatisfiesTemplate() bool {
+func (m *manager) certSatisfiesTemplate() bool {
 	m.certAccessLock.RLock()
 	defer m.certAccessLock.RUnlock()
 	if m.cert == nil {

Run ./hack/update-gofmt.sh
```

liggitt commented on staging/src/k8s.io/client-go/util/certificate/certificate_manager.go:
```diff
@@ -389,35 +389,30 @@ func (m *manager) rotateCerts() (bool, error) {
-			return time.Now()
+			glog.V(2).Infof("Current certificate CN (%s) does not match requested CN (%s)", m.cert.Leaf.Subject.CommonName, template.Subject.CommonName)
+			return false
 		}
```
also check `sets.NewString(m.cert.Leaf.Subject.Organizations...).HasAll(template.Subject.Organizations...)`

liggitt commented:
/test pull-kubernetes-e2e-gke

liggitt commented:
thanks for the PR, just a couple comments. exercising the logic in a unit test would be good as well

agunnerson-ibm commented:
Thanks everyone for the comments! I've updated this PR to address them.

Changes:
* Fixed all gofmt errors
* Added check to ensure that the organizations in the x509 subject of the cert match the template
* Split `certSatisfiesTemplate` into `certSatisfiesTemplate` and `certSatisfiesTemplateLocked` and use the latter if `certAccessLock` is already locked
* Added unit tests for `certSatisfiesTemplate`

agunnerson-ibm commented:
Fixed the lint issue reported by `pull-kubernetes-verify`.

awly commented on staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go:
```diff
@@ -212,6 +213,178 @@ func TestSetRotationDeadline(t *testing.T) {
+			}
+
+			if m.certSatisfiesTemplate() != tc.shouldSatisfy {
+				if tc.shouldSatisfy {
```
Don't wrap errors, store `certSatisfiesTemplate` result in a var and do a simpler error print:
```
t.Errorf("cert: %+v, template: %+v, certSatisfiesTemplate returned %v, want %v", m.cert, tc.template, got, tc.shouldSatisfy)
```

agunnerson-ibm replied:
Thanks, I'll fix that.

awly commented on staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go:
```diff
@@ -212,6 +213,178 @@ func TestSetRotationDeadline(t *testing.T) {
+			m := manager{
+				cert:        tlsCert,
+				getTemplate: func() *x509.CertificateRequest { return tc.template },
+				usages:      []certificates.KeyUsage{},
```
Is this needed?

agunnerson-ibm replied:
Nope, I'll remove it.

awly reviewed (commented):
Two small nits, LGTM otherwise

agunnerson-ibm commented:
The latest commit should address @awly's comments above.
